pkg/fanal/image/daemon/image.go

```diff
@@ -153,12 +154,39 @@ func (img *image) configHistory() []v1.History {
 			},
 			CreatedBy:  h.CreatedBy,
 			Comment:    h.Comment,
-			EmptyLayer: h.Size == 0,
+			EmptyLayer: emptyLayer(h),
 		})
 	}
 	return history
 }
 
+func emptyLayer(history dimage.HistoryResponseItem) bool {
+	if history.Size != 0 {
+		return false
+	}
+	createdBy := strings.TrimSpace(strings.TrimLeft(history.CreatedBy, "/bin/sh -c #(nop)"))
+	// This logic is taken from https://github.com/moby/buildkit/blob/2942d13ff489a2a49082c99e6104517e357e53ad/frontend/dockerfile/dockerfile2llb/convert.go
+	if strings.HasPrefix(createdBy, "ENV") ||
+		strings.HasPrefix(createdBy, "MAINTAINER") ||
+		strings.HasPrefix(createdBy, "LABEL") ||
+		strings.HasPrefix(createdBy, "CMD") ||
+		strings.HasPrefix(createdBy, "ENTRYPOINT") ||
+		strings.HasPrefix(createdBy, "HEALTHCHECK") ||
+		strings.HasPrefix(createdBy, "EXPOSE") ||
+		strings.HasPrefix(createdBy, "USER") ||
+		strings.HasPrefix(createdBy, "VOLUME") ||
+		strings.HasPrefix(createdBy, "STOPSIGNAL") ||
+		strings.HasPrefix(createdBy, "SHELL") ||
+		strings.HasPrefix(createdBy, "ARG") ||
+		createdBy == "WORKDIR /" { // only when workdir == "/" then layer is empty
+		return true
+	}
+	// commands here: 'ADD', COPY, RUN and WORKDIR != "/"
+	// Also RUN command may don't include 'RUN' prefix
```

Comment:
Does it happen? I thought `RUN` was required.

Reply:
```suggestion
	// Also RUN command may not include 'RUN' prefix
```

Reply:
Yes. Image created with `docker build` command may not include command prefix.
e.g. 
Dockerfile:
```dockerfile
FROM alpine

RUN mkdir test
```
History:
```zsh
➜ docker history 2780:test
IMAGE          CREATED        CREATED BY                                      SIZE      COMMENT
4d55786712ed   8 hours ago    /bin/sh -c mkdir test                           0B        
0ac33e5f5afa   4 months ago   /bin/sh -c #(nop)  CMD ["/bin/sh"]              0B        
<missing>      4 months ago   /bin/sh -c #(nop) ADD file:5d673d25da3a14ce1…   5.57MB    
```

Also layers created with `buildkit` have command prefix:
```zsh
➜ docker history 2780:test2                                      
IMAGE          CREATED         CREATED BY                                      SIZE      COMMENT
d3fa167c3c1e   3 seconds ago   RUN /bin/sh -c mkdir test # buildkit            0B        buildkit.dockerfile.v0
<missing>      4 months ago    /bin/sh -c #(nop)  CMD ["/bin/sh"]              0B        
<missing>      4 months ago    /bin/sh -c #(nop) ADD file:5d673d25da3a14ce1…   5.57MB 
```